File: Kalorie/Kalorie/Core/UseCases/FetchMaintainerClaimUseCase.swift
//
//  FetchMaintainerClaimUseCase.swift
//  Kalorie
//
//  Created by Josef Antoni on 10.09.2026.
//

import FirebaseAuth
import Foundation

protocol FetchMaintainerClaimUseCaseProtocol {
    func callAsFunction() async throws -> Bool
}

final class MaintainerClaimCache {

    // MARK: - Properties

    var value: Bool?
    var cachedAt: Date?

    // MARK: - Init

    init() {}
}

struct FetchMaintainerClaimUseCase: FetchMaintainerClaimUseCaseProtocol {

    // MARK: - Properties

    private let cache: MaintainerClaimCache

    // MARK: - Init

    init(cache: MaintainerClaimCache) {
        self.cache = cache
    }

    // MARK: - Functions

    func callAsFunction() async throws -> Bool {
        if
            let cachedValue = cache.value,
            let cachedAt = cache.cachedAt,
            Date.now.timeIntervalSince(cachedAt) < Constants.Auth.maintainerClaimCacheTTL
        {
            return cachedValue
        }
        guard let user = Auth.auth().currentUser else { throw AuthError.notAuthenticated }
        let result = try await user.getIDTokenResult(forcingRefresh: cache.value != nil)
        let isMaintainer = result.claims["maintainer"] as? Bool ?? false
        cache.value = isMaintainer
        cache.cachedAt = .now
        return isMaintainer
    }
}

#if DEBUG
struct FetchMaintainerClaimUseCaseFake: FetchMaintainerClaimUseCaseProtocol {

    // MARK: - Properties

    var stubbedIsMaintainer = false

    // MARK: - Functions

    func callAsFunction() async throws -> Bool {
        stubbedIsMaintainer
    }
}
#endif

File: Android/app/src/main/java/antoni/kalorie/core/usecases/FetchMaintainerClaimUseCase.kt
package antoni.kalorie.core.usecases

import antoni.kalorie.core.auth.AuthError
import antoni.kalorie.core.utils.Constants
import com.google.firebase.auth.FirebaseAuth
import java.time.Duration
import java.time.Instant
import kotlinx.coroutines.tasks.await

interface FetchMaintainerClaimUseCaseProtocol {
    suspend operator fun invoke(): Boolean
}

class MaintainerClaimCache {

    // MARK: - Properties

    data class Entry(val value: Boolean, val cachedAt: Instant)

    var entry: Entry? = null
}

class FetchMaintainerClaimUseCase(
    private val cache: MaintainerClaimCache,
    private val auth: FirebaseAuth = FirebaseAuth.getInstance(),
) : FetchMaintainerClaimUseCaseProtocol {

    // MARK: - Functions

    override suspend fun invoke(): Boolean {
        val cached = cache.entry
        if (cached != null && Duration.between(cached.cachedAt, Instant.now()).seconds < Constants.Auth.MAINTAINER_CLAIM_CACHE_TTL_SECONDS) {
            return cached.value
        }
        val user = auth.currentUser ?: throw AuthError.NotAuthenticated
        val result = user.getIdToken(cache.entry != null).await()
        val isMaintainer = result.claims["maintainer"] as? Boolean ?: false
        cache.entry = MaintainerClaimCache.Entry(value = isMaintainer, cachedAt = Instant.now())
        return isMaintainer
    }
}
